cleverhans/attacks.py

```diff
@@ -749,6 +749,91 @@ def parse_params(self, y=None, y_target=None, nb_classes=None,
         self.clip_max = clip_max
 
 
+class DeepFool(Attack):
+
+    """
+    DeepFool is an untargeted & iterative attack which aims at finding the
+    minimum adversarial perturbations in deep networks. The implementation
+    here is w.r.t. the L2 norm.
+    Paper link: "https://arxiv.org/pdf/1511.04599.pdf"
+    """
+
+    def __init__(self, model, back='tf', sess=None):
+        """
+        Create a DeepFool instance.
+        """
+        super(DeepFool, self).__init__(model, back, sess)
+
+        if self.back == 'th':
+            raise NotImplementedError('Theano version not implemented.')
+
+        import tensorflow as tf
+        self.structural_kwargs = ['over_shoot', 'max_iter',
+                                  'clip_max', 'clip_min',
+                                  'nb_candidate', 'nb_classes']
+
+        if not isinstance(self.model, Model):
+            self.model = CallableModelWrapper(self.model, 'logits')
+
+    def generate(self, x, **kwargs):
+        """
+        Generate symbolic graph for adversarial examples and return.
+        :param x: The model's symbolic inputs.
+        :param nb_candidate: The number of classes to test against, i.e.,
+                            deepfool only consider nb_candidate classes when
+                            attacking (thus accelerate speed)
+        :param overshoot: A termination criterion to prevent vanishing updates
+        :param max_iter: Maximum number of iteration for deepfool
+        :param nb_classes: The number of model output classes
+        :param clip_min: Minimum component value for clipping
+        :param clip_max: Maximum component value for clipping
+        """
+
+        import tensorflow as tf
+        from .attacks_tf import gradient_graph, deepfool_batch
+
+        # Parse and save attack-specific parameters
+        assert self.parse_params(**kwargs)
+
+        assert self.nb_candidate <= self.nb_classes,\
+            'nb_candidate should not be greater than nb_classes'
+
+        # Define graph wrt to this input placeholder
+        logits = self.model.get_logits(x)
+        preds = tf.reshape(tf.nn.top_k(logits, k=self.nb_candidate)[0],
+                           [-1, self.nb_candidate])
+        grads = gradient_graph(preds, x, self.nb_candidate)
+
+        # Define graph
+        def deepfool_wrap(x_val):
+            return deepfool_batch(self.sess, x, preds, logits, grads, x_val,
+                                  self.nb_candidate, self.overshoot,
+                                  self.max_iter, self.clip_min, self.clip_max,
+                                  self.nb_classes)
+        return tf.py_func(deepfool_wrap, [x], tf.float32)
+
+    def parse_params(self, nb_candidate=10, overshoot=0.02, max_iter=50,
+                     nb_classes=1001, clip_min=0., clip_max=1., **kwargs):
```

Comment:
Can you infer the number of classes from the shape of `preds` tensor instead of feeding it as an argument to this function?

Reply:
Yes. We have changed that. Thanks for your suggestion